### PR TITLE
CPM-624 add SetTableValue user intent

### PR DESCRIPTION
**Pull request by amigaire**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [x] Tests

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/SetTableValue.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class SetTableValue implements ValueUserIntent
+{
+    public function __construct(
+        private string $attributeCode,
+        private ?string $channelCode,
+        private ?string $localeCode,
+        private array $tableValue
```
Comment (cemorin) [inline src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/SetTableValue.php:19]: Do we allow empty arrays?

**Review by cemorin — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/SetTableValue.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+use Webmozart\Assert\Assert;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class SetTableValue implements ValueUserIntent
+{
+    public function __construct(
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/SetTableValue.php:15]:
```suggestion
    /**
     * @param array<int, array<string, mixed>> $tableValue
     */
    public function __construct(
```

To explain what is expected. Don't forget the user intents are created for people that possibly does not know our bounded context.
You can also add an example if you want.

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php [hunk 3]
```diff
@@ -1107,6 +1110,144 @@ public function it_throws_an_exception_when_currency_does_not_exist(): void
         ]));
     }
 
+    /** @test */
+    public function it_creates_and_updates_a_product_with_a_table_value(): void
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php:1114]: This test should only be ran if the table attribute feature is enabled

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php [hunk 4]
```diff
@@ -1107,6 +1110,144 @@ public function it_throws_an_exception_when_currency_does_not_exist(): void
...
+                        ['ingredient' => 'egg', 'quantity' => 2],
+                    ]
+                )
+            ]
+        );
+        $this->messageBus->dispatch($command);
+
+        $this->clearDoctrineUoW();
+
+        $tableNormalizer = new TableNormalizer();
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php:1164]:
```suggestion
        $tableNormalizer = $this->get('...');
```

Using the service is better for the future, if we nned to inject something in the class for instance

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php [hunk 5]
```diff
@@ -1107,6 +1110,144 @@ public function it_throws_an_exception_when_currency_does_not_exist(): void
...
+                    'a_table',
+                    null,
+                    null,
+                    ['ingredient' => 'butter', 'quantity' => 3]
+                )
+            ]
+        );
+        $this->messageBus->dispatch($command);
+
+        $this->clearDoctrineUoW();
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php:1248]:
```suggestion
```

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php [hunk 6]
```diff
@@ -1107,6 +1110,144 @@ public function it_throws_an_exception_when_currency_does_not_exist(): void
...
+            [
+                ['ingredient' => 'butter', 'quantity' => 3],
+                ['ingredient' => 'egg'],
+            ],
+            $tableNormalizer->normalize($product->getValue('a_table')->getData())
+        );
+    }
+
+    /** @test */
+    public function it_throws_an_exception_when_table_value_has_wrong_format(): void // todo naming ?????
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php:1200]: Ok for the naming
But should be ran only if feature is activated

File: src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/SetTableValue.php (hunk 1 above)
Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/SetTableValue.php:19]: Yes, we allow empty values in SET use intents

**Review by nmarniesse — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/SetTableValue.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+     *             1 => ['ingredient' => 'salt'],
+     *         ]
+     */
+    public function __construct(
+        private string $attributeCode,
+        private ?string $channelCode,
+        private ?string $localeCode,
+        private array $tableValue
+    ) {
+        Assert::allIsArray($tableValue);
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/API/Command/UserIntent/SetTableValue.php:28]:
```suggestion
        Assert::isList($tableValue);
        Assert::allIsMap($tableValue);
```
(this way you can validate that the keys of the rows are integers, and the keys of the cells are strings, so the phpDoc is always correct

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/API/Command/UserIntent/SetTableValueSpec.php [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+    function it_returns_the_table_value()
+    {
+        $this->tableValue()->shouldBeLike(
+            [
+                ['ingredient' => 'salt'],
+                ['ingredient' => 'egg', 'quantity' => 2],
+            ]
+        );
+    }
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/API/Command/UserIntent/SetTableValueSpec.php:59]: can you add a test like 'it_must_be_instantiated_with_valid_data_structure' or smth like that? it should check that the $tableValue structure is correct

**Review by mmetayer — APPROVED** [review]

**Pull request merged**